packages/nimble-components/src/rich-text/specs/mention-hld.md

```diff
@@ -0,0 +1,725 @@
+
+The `mention-update` event will also be triggered when the user types any character after `@`, containing that text along with the `@` character
+in the event data. For example, if a user types `@` and then adds `a` the event will be emitted with data that includes the value `@a`.
+The client can listen to this event, filter the list of users that includes the names containing the letter `a`, and then dynamically
+update the `nimble-mapping-mention-user` element based on the filter data. Subsequently, a maximum of 20 filtered options should be
+transmitted to the editor.
+
+Since the above event triggers for every key down event like adding/removing texts, move the text cursors after the `@` character which is quite an
+expensive operation to perform for every keystroke so it is advisable to `debounce` the events if you're using network requests to perform the
+filtering operations. Like, allow at most one request per second to filter the list for each second instead of for each keystrokes.
+
+_Note_: The editor will also perform the same filtering once again to ensure the filtered options are proper and update the dropdown list in the UI.
+This helps to filter the list, regardless of whether the client is loading the list dynamically by listening to the event as mentioned above
+or statically providing user details at the start via `nimble-mapping-mention`.
+
+### API
+
+To support the `@mention`, the creation of the following configuration elements and UI elements is required for rich text components:
+
+#### User mention element (Non-visible configuration element):
+
+This element serves as a container for the mapping elements and additionally functions as a configuration element. It allows the transmission of
+arbitrary values that are specific to the `@mention` users within the rich text components. Currently, this component has the below
+configuration properties/attributes for user mentions and plays the role of containing the user mapping elements.
+
+This component can be used as a feature flag to enable the support for `@mention` in the rich text editor and viewer.
+If this component is not present at the children of the rich text editor or viewer, the component will function the same without the support
+for `@mention` in any place like editor, viewer, markdown parser and serializer and the toolbar button to add **"@"** in the editor
+will be hidden.
+
+_Component Name_
+
+-   `nimble-rich-text-mention-users`
+
+_Content_
+
+-   Zero or more `nimble-mapping-mention-user` elements
+
+_Props/Attrs_
+
+-   `pattern` - is a string attribute that gets the regex pattern to match the `mention-href` in the mapping element. Exactly similar to
+    HTML attribute: [pattern](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/pattern).
+-   `mentioned-hrefs` - is a read-only property that returns an array of strings representing the mentioned user URL (`mention-href` property value of
```

Comment:
How does a client know when `mentioned-hrefs` changes? The `mention-update` only fires when a user is interactively entering a mention using the dropdown list workflow. But what about if the user copy and pastes some text? Should the client listen for all `input` events and then check the value of `mentioned-hrefs` for changes?

Reply:
> But what about if the user copy and pastes some text?

Maybe I'm not following but how does your comment address that workflow (copy and paste some text)? The client will not create mapping elements for all users at the company. The client needs to know when the list of mentioned-users changes so that new mapping elements can be created as needed right?

Reply:
Actually, when we copy a mention node, it copies along with the URL and a display name. So when we paste the same into the editor, it renders as the mention node as the URL and name are preserved in the element itself and not require to have a new mapping element for the copied users from the client.

Here is the short gif (Copied mention node from the viewer and pasted into the editor that does not have the mapping element for the copied user): 
![copy-paste-mention](https://github.com/ni/nimble/assets/123377523/afaa42a6-7289-4de3-aca2-a7fe63933989)

Both copied from (viewer) and pasted into (editor) will have the same element in DOM with the attribute values preserved.
![image](https://github.com/ni/nimble/assets/123377523/348582b4-91d0-4f5c-94d4-ebe100a047ef)

Reply:
That conceptually seems wrong. If you paste in a mention then add a new mention, what will make it get populated again? If the configuration changes (new mentions are added) and you call getMarkdown and setMarkdown what will make it render the same?

This goes back to the declarative definition concept. What your showing does not represent the state of the component correctly. The only mentions that should be resolved are those that have mappings defined. The paste operation is not behaving well because it's letting outside state into the system.

Reply:
@vivinkrishna-ni FYI this thread was not resolved (I know y'all were looking to try and resolve the HLD feedback more quickly)

Reply:
> The paste operation is not behaving well because it's letting outside state into the system.

In that case, we might have to do emit the `mention-update` event whenever a mention node is pasted into the editor. This event will be fired only after pushing the user URL(s) from the pasted content to `mentioned-hrefs` array. The client can listen to the event, read the mentioned Href value, and send the corresponding mapping element(s). 

At first, only the href will be pasted into the editor as links which can be achieved with the [`transformPasted`](https://prosemirror.net/docs/ref/#view.EditorProps.transformPasted) (confirmed it with a prototype). Once the client sends the new mapping element, we will call getMarkdown and setMarkdown which will update the link to a mention node. If the mapping element is not given, then it stays as a link.

I have updated the HLD about the behavior and client usage guidance section to address pasting behavior.